### PR TITLE
feat: add transaction type

### PR DESCRIPTION
**Pull request by bajpai244**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

Resolves: #654 #651 

## What is the new behavior?

- adds a new type for Ethereum Transaction
- adds the logic to parse AccessLists
- makes `compute_rlp_encoding.ts` easier to use

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by enitrat — CHANGES_REQUESTED** [review]
> Reviewed 18 of 18 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/659)* status: all files reviewed, 14 unresolved discussions (waiting on @bajpai244 and @ClementWalter)

___
*[`log.txt` line 0 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/659#-NlYA2L21iUuz-lC6EIE:-NlYA2L21iUuz-lC6EIF:btcc0ao) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/a1370c25b843500f7424b067f62a564179fba678/log.txt#L0)):*
What is this file?
___
*[`crates/contracts/src/kakarot_core/kakarot.cairo` line 41 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/659#-NlYAEKT0G0CCq7eqvSO:-NlYAEKT0G0CCq7eqvSP:bfyyhn9) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/a1370c25b843500f7424b067f62a564179fba678/crates/contracts/src/kakarot_core/kakarot.cairo#L41)):*
> ```Cairo
>     use super::{StoredAccountType};
>     use utils::address::compute_contract_address;
>     use utils::checked_math::checked_math::CheckedMath;
> ```

keep import from external module - which reexports CheckedMath
___
*[`crates/contracts/src/kakarot_core/kakarot.cairo` line 365 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/659#-NlYAOz0-O0gMkKnhSzl:-NlYAOz0-O0gMkKnhSzm:b-si3fhd) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/a1370c25b843500f7424b067f62a564179fba678/crates/contracts/src/kakarot_core/kakarot.cairo#L365)):*
> ```Cairo
>             let value = tx.value();
>             let to = tx.destination();
>             let calldata = tx.calldata();
> ```

We can remove these, they don't bring any clarity nor performance gains - especially given that they're not being reused



_Code quote:_
```Cairo
            let gas_price = tx.gas_price();
            let gas_limit = tx.gas_limit();
            let value = tx.value();
            let to = tx.destination();
            let calldata = tx.calldata();
```
___
*[`crates/contracts/src/kakarot_core/kakarot.cairo` line 459 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/659#-NlYAoxpAwQK2pC4l7ni:-NlYAoxpAwQK2pC4l7nj:bjlhzk4) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/a1370c25b843500f7424b067f62a564179fba678/crates/contracts/src/kakarot_core/kakarot.cairo#L459)):*
> ```Cairo
>                 caller: origin,
>                 target: to,
>                 gas_limit: tx.gas_limit(),
> ```

no, this is the gas_left after having charged the intrinsic gas costs
___
*[`crates/evm/src/gas.cairo` line 133 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/659#-NlYELrYA7NERiHVAmlQ:-NlYELrYA7NERiHVAmlR:b-ttmk9z) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/a1370c25b843500f7424b067f62a564179fba678/crates/evm/src/gas.cairo#L133)):*
> ```Cairo
>     let mut access_list_cost: u128 = 0;
> 
>     let mut calldata_cpy: Span<u8> = calldata;
> ```

what is this for?
___
*[`crates/utils/src/eth_transaction.cairo` line 15 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/659#-NlYEVhA4qfqxvPhe9kz:-NlYEVhA4qfqxvPhe9l-:bhtisc9) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/a1370c25b843500f7424b067f62a564179fba678/crates/utils/src/eth_transaction.cairo#L15)):*
> ```Cairo
> use utils::rlp::{RLPTrait, RLPHelpersTrait};
> 
> #[derive(Copy, Clone, Drop, Serde, PartialEq, Display)]
> ```

derive Debug instead of Display? 
___
*[`crates/utils/src/eth_transaction.cairo` line 91 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/659#-NlYHmmJ9lOXOifa2NPg:-NlYHmmJ9lOXOifa2NPh:b-jhfdp2) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/a1370c25b843500f7424b067f62a564179fba678/crates/utils/src/eth_transaction.cairo#L91)):*
> ```Cairo
>     }
> 
>     fn value(self: @EthereumTransaction) -> u256 {
> ```

are these used? If not, just keep the ones used
___
*[`crates/utils/src/eth_transaction.cairo` line 289 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/659#-NlYEsau0lemxMh6dFNR:-NlYEsau0lemxMh6dFNS:b-6143j) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/a1370c25b843500f7424b067f62a564179fba678/crates/utils/src/eth_transaction.cairo#L289)):*
> ```Cairo
> 
>         let decoded_data = *decoded_data.at(0);
>         let decoded_data = match decoded_data {
> ```

merge these two lines
___
*[`crates/utils/src/rlp.cairo` line 261 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/659#-NlYH7MnCZ5Lw6q6FJwe:-NlYH7MnCZ5Lw6q6FJwf:b-iecdog) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/a1370c25b843500f7424b067f62a564179fba678/crates/utils/src/rlp.cairo#L261)):*
> ```Cairo
>     }
> 
>     fn parse_access_list(self: RLPItem) -> Result<Span<AccessListItem>, RLPHelpersError> {
> ```

overall this function needs to be refactored:
Each internal loop needs to be extracted into its own function, e.g. 
`extract_accesses_addresses`, `extract_accessed_keys`, `extract_access_list` (or similar)
___
*[`crates/utils/src/rlp.cairo` line 264 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/659#-NlYFjpB2yI4xMb4rfwS:-NlYFjpB2yI4xMb4rfwT:b-aqcxxy) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/a1370c25b843500f7424b067f62a564179fba678/crates/utils/src/rlp.cairo#L264)):*
> ```Cairo
>         match self {
>             RLPItem::String(_) => { Result::Err(RLPHelpersError::NotAList) },
>             RLPItem::List(mut v) => {
> ```

what is `v` ?
___
*[`crates/utils/src/rlp.cairo` line 267 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/659#-NlYFxhy5MlLcmHeAs_N:-NlYFxhy5MlLcmHeAs_O:b-1pe4p2) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/a1370c25b843500f7424b067f62a564179fba678/crates/utils/src/rlp.cairo#L267)):*
> ```Cairo
>                 let res: Result<Span<AccessListItem>, RLPHelpersError> = loop {
>                     let mut access_list: Array<AccessListItem> = array![];
>                     let list = match v.pop_front() {
> ```

is this actually an `AccessItem`? in which case renaming it `item` would be better
___
*[`crates/utils/src/rlp.cairo` line 286 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/659#-NlYGHIQ4Oc10oZ_m80Z:-NlYGHIQ4Oc10oZ_m80_:b4fg7f8) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/a1370c25b843500f7424b067f62a564179fba678/crates/utils/src/rlp.cairo#L286)):*
> ```Cairo
> 
>                     let ethereum_address = match ((*list.at(0)).try_parse_address_from_string()) {
>                         Result::Ok(v) => {
> ```

give meaningful names, here `maybe_address` is a lot easier to understand than just `v`
___
*[`crates/utils/src/rlp.cairo` line 299 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/659#-NlYGxaa2ZT7EhOFcM3f:-NlYGxaa2ZT7EhOFcM3g:b5a56gd) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/a1370c25b843500f7424b067f62a564179fba678/crates/utils/src/rlp.cairo#L299)):*
> ```Cairo
>                     let storage_keys: Span<u256> = match (*list.at(1)) {
>                         RLPItem::String(_) => { break Result::Err(RLPHelpersError::NotAList); },
>                         RLPItem::List(mut list) => {
> ```

if what you're getting is a list of keys, name it `keys`



_Code quote:_
```Cairo
RLPItem::List(mut list) => {
```
___
*[`crates/utils/src/tests/test_eth_transaction.cairo` line 31 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/659#-NlYHVEb1yL3RDKiGgEp:-NlYHVEb1yL3RDKiGgEq:b-ttxi41) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/a1370c25b843500f7424b067f62a564179fba678/crates/utils/src/tests/test_eth_transaction.cairo#L31)):*
> ```Cairo
>     let tx = encoded_tx.decode().expect('decode failed');
>     let tx = match (tx) {
>         EthereumTransaction::LegacyTransaction(v) => { v },
> ```

same here, for good habits: `tx` is better than just `v`

**Review by enitrat — CHANGES_REQUESTED** [review]
> Reviewed 3 of 6 files at r2, 1 of 1 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/659)* status: 17 of 18 files reviewed, 2 unresolved discussions (waiting on @bajpai244 and @ClementWalter)

___
*[`crates/utils/src/eth_transaction.cairo` line 15 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/659#-NlYEVhA4qfqxvPhe9kz:-NlZ88KF6YPpv4H_5iGu:bhplx4p) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/a1370c25b843500f7424b067f62a564179fba678/crates/utils/src/eth_transaction.cairo#L15)):*
Hm, I think in this case it should derive Debug, not Display, because we don't actually want to display the values: the only reason why we would use it is for debugging.

From the Rust book:
> Yeah, `Debug` is simple and easy to use. But sometimes we want to customize the output appearance of our type. This is where `Display` really shines.
>Unlike `Debug`, there is no way to derive the implementation of the `Display` trait, we have to manually implement it.

Basically - Display is to display in a well formatted manner, while debug is only for raw formatting.

And Span/Array implement Debug, not Display

**Review by bajpai244 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/659)* status: 16 of 19 files reviewed, 2 unresolved discussions (waiting on @ClementWalter and @enitrat)

___
*[`crates/utils/src/eth_transaction.cairo` line 15 at r1](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/659#-NlYEVhA4qfqxvPhe9kz:-NlZCTE37MTghtkQjzXc:b-3j4joq) ([raw file](https://github.com/kkrt-labs/kakarot-ssj/blob/a1370c25b843500f7424b067f62a564179fba678/crates/utils/src/eth_transaction.cairo#L15)):*
<details><summary><i>Previously, enitrat (Mathieu) wrote…</i></summary><blockquote>

Hm, I think in this case it should derive Debug, not Display, because we don't actually want to display the values: the only reason why we would use it is for debugging.

From the Rust book:
> Yeah, `Debug` is simple and easy to use. But sometimes we want to customize the output appearance of our type. This is where `Display` really shines.
>Unlike `Debug`, there is no way to derive the implementation of the `Display` trait, we have to manually implement it.

Basically - Display is to display in a well formatted manner, while debug is only for raw formatting.

And Span/Array implement Debug, not Display
</blockquote></details>

Gotcha, done!

**Review by enitrat — APPROVED** [review]
